qubes/ext/gui.py

```diff
@@ -29,17 +29,45 @@ class GUI(qubes.ext.Extension):
     # TODO put this somewhere...
     @staticmethod
     def send_gui_mode(vm):
-        vm.run_service('qubes.SetGuiMode',
-            input=('SEAMLESS'
-            if vm.features.get('gui-seamless', False)
+        vm.run_service('qubes.SetGuiMode', input=(
+            'SEAMLESS' if vm.features.get('gui-seamless', False)
             else 'FULLSCREEN'))
 
+    @qubes.ext.handler('property-set:guivm')
+    def on_property_set(self, subject, event, name, newvalue, oldvalue=None):
+        # Clean other 'guivm-XXX' tags.
+        # gui-daemon can connect to only one domain
+        tags_list = list(subject.tags)
+        for tag in tags_list:
+            if 'guivm-' in tag:
+                subject.tags.remove(tag)
+
+        guivm = 'guivm-' + newvalue.name
+        subject.tags.add(guivm)
+
     @qubes.ext.handler('domain-qdb-create')
     def on_domain_qdb_create(self, vm, event):
         # pylint: disable=unused-argument,no-self-use
         for feature in ('gui-videoram-overhead', 'gui-videoram-min'):
             try:
-                vm.untrusted_qdb.write('/qubes-{}'.format(feature),
+                vm.untrusted_qdb.write(
+                    '/qubes-{}'.format(feature),
                     vm.features.check_with_template_and_adminvm(feature))
             except KeyError:
                 pass
+
+        # Add GuiVM Xen ID for gui-daemon
+        try:
+            if vm.guivm is not None:
+                if str(vm.name) != str(vm.guivm.name):
+                    vm.untrusted_qdb.write('/qubes-gui-domain-xid',
+                                           str(vm.guivm.xid))
+        except AttributeError:
+            vm.untrusted_qdb.write('/qubes-gui-domain-xid', '')
+
+        # Add keyboard layout from that of GuiVM
+        try:
+            kbd_layout = vm.guivm.features['kbd_layout']
```

Comment:
Generally we use '-' as word separator in tags, features etc. And also it would be more readable to use full word (`keyboard-layout`).